docs/proposals/drafts/20211106-plugable-authorizers.md

```diff
@@ -0,0 +1,97 @@
+---
+toc: true
+title: 2021-11 Pluggable Authorizers
+---
+
+## Introduce Pluggable Authorizers
+
+* **Owners**
+  * @tareqmamari
+
+## TL; DR
+
+We propose a pluggable authorizers support to enable plugging in custom authorizers in a self-register manner besides the already supported authorizers in Observatorium API.
+
+## Why
+
+This proposed approach allows the adopters of Observatorium to introduce and use their own custom in-process authorizers' implementation. Currently, adopters can build their own OPA integration, which means introducing another component where OPA server/endpoints are running. In one hand, it allows adopters to use their own custom implementation, however, it also comes with additional efforts such as managing the new OPA-integration component's lifecycle. There are adopters, who prefer to have custom in-process authorizers other than using OPA-integration for different reasons, one important reason besides avoid managing a new component, in-process authorizers tend to work better for medium to large workloads.
+
+## Goals
+
+- Do not change the current common interface for the authorizers.
+- Provide a self-registering mechanism of the authorizers so that the adopters do not need to patch the current codebase to introduce their own custom authorization logic.
+- The new changes must be backward-compatible. In other words, the current authorizer implementation and configuration must not be broken by this change and should be supported to use the new pluggable based implementation.
+
+## How
+
+We propose to extend rego evaluation logic to load custom built-in functions in Go, in this way, adopters can implement their own authorization logic in go and then call it in Rego. In addition, we aim to provide a self-registering mechanism, so that introducing any new custom built-in function does not necessary require to do any patch/change in the existing codebase, instead, only the plugin's go source file is needed other than patching the other source files to add the plugin. For that, we propose to use importing for side-effect mechanism, where a protected map of authorizer's provider type and their custom built-in function can be used for such scenario.
+
+## Implementation Options
+1. Custom Rego Built-in Functions in Go
+2. Importing for Side-Effect
+3. Go Plugin Module
+
+## 1. Custom Rego Built-in Functions in Go
+
+Instead of totally introducing a new authorizer, OPA inprocess authorizer ( using Rego ) can be extended through [custom rego built-in functions in go](https://www.openpolicyagent.org/docs/latest/extensions/#custom-built-in-functions-in-go). In this approach, custom auhtorization go logic can be introduced as a custom built-in rego function in go, which then can be called in the rego logic. Once rego calls this function, the go implementation of it, would be executed. To make it pluggable, importing for side-effects is used, similar to what is proposed in the second option.
```

Comment:
```suggestion
Instead of totally introducing a new authorizer, OPA inprocess authorizer (using Rego) can be extended through [custom Rego built-in functions in go](https://www.openpolicyagent.org/docs/latest/extensions/#custom-built-in-functions-in-go). In this approach, custom auhtorization go logic can be introduced as a custom built-in Rego function in go, which then can be called in the rego logic. Once Rego calls this function, the go implementation of it, would be executed. To make it pluggable, importing for side-effects is used, similar to what is proposed in the second option.
```